### PR TITLE
Fix: improve links performance

### PR DESCRIPTION
**Pull request by madcampos**
Try to improve link performance

### CODE REVIEW

File: src/data/StateTypes.ts [hunk 1]
```diff
@@ -75,7 +70,7 @@ export interface LinkGraphState {
 
 export interface NetworkCongestionState {
     linkSaturationTreshold: number;
-    graphs: Record<string, LinkGraphState>;
+    graphs: { links: Record<string, LinkState>; totalOps: number }[];
```
Comment (aidemsined) [inline src/data/StateTypes.ts:73]: this is the variable we really need to rename even in draft

File: src/data/GraphOnChip.ts [hunk 2]
```diff
@@ -1067,12 +1100,19 @@ export class ComputeNode {
...
 
-        const linkId = `${node.loc.x}-${node.loc.y}`;
+        const linkId = node.uid;
 
         Object.entries(nodeJSON.links).forEach(([linkName, linkJson], index) => {
             const link: NetworkLink = NetworkLink.CREATE(linkName as NOCLinkName, `${linkId}-${index}`, linkJson);
             if (link.type === LinkType.NOC) {
-                node.links.set(linkName, link as NOCLink);
+                // Added a const here to avoid casting multiple times
+                const NocLink = link as NOCLink;
```
Comment (aidemsined) [inline src/data/GraphOnChip.ts:1109]: capitalization

File: src/renderer/components/NodeGridElement.tsx [hunk 3]
```diff
@@ -48,8 +50,8 @@ const NodeGridElement: React.FC<NodeGridElementProps> = ({ node, graphName, temp
     const isHighContrast = useSelector(getHighContrastState);
     const showLinkSaturation = useSelector(getShowLinkSaturation);
 
-    // TODO: pre-calculate link saturation and memoize it
-    const linksData = useSelector(getAllLinksForGraph(graphName));
+    const linksData = useSelector(getNodeLinksData(temporalEpoch, node.uid));
+    const nodeSaturation = useSelector(getOffchipNodeSaturation(temporalEpoch, node.uid));
```
Comment (aidemsined) [inline src/renderer/components/NodeGridElement.tsx:54]: maybe `offchipLinkSaturation`  and `offchipLinkSaturationForNode` as a selector?

File: src/renderer/components/node-grid-elements-components/OffChipNodeLinkCongestionLayer.tsx [hunk 4]
```diff
@@ -21,45 +17,12 @@ interface OffChipNodeLinkCongestionLayerProps {
  * This renders a congestion layer for nodes with off chip links (DRAM, Ethernet, PCIe)  for those links
  */
 const OffChipNodeLinkCongestionLayer: FC<OffChipNodeLinkCongestionLayerProps> = ({
-    node,
-    linksData,
+    saturation,
```
Comment (aidemsined) [inline src/renderer/components/node-grid-elements-components/OffChipNodeLinkCongestionLayer.tsx:20]: offchip reference?

File: src/renderer/components/node-grid-elements-components/OffChipNodeLinkCongestionLayer.tsx [hunk 5]
```diff
@@ -21,45 +17,12 @@ interface OffChipNodeLinkCongestionLayerProps {
...
     showLinkSaturation,
     isHighContrast,
 }) => {
     const linkSaturationTreshold = useSelector(getLinkSaturation);
```
Comment (aidemsined) [inline src/renderer/components/node-grid-elements-components/OffChipNodeLinkCongestionLayer.tsx:24]: migh as well fix the spelling `linkSaturationThreshold`

**Comment by aidemsined** [comment]
```getGraphOnChip: (
        temporalEpoch: number,
        chipId?: number,
    ) => { graph: GraphOnChip; relationship: GraphRelationship }[];```
    this method shoudl be returned to itrs original state. 
`getGraphOnChipListForTemporalEpoch` should be used instead to return an array which is index by `chipId`

File: src/data/GraphOnChip.ts [hunk 6]
```diff
@@ -757,22 +758,93 @@ export default class GraphOnChip {
         return [...nodes];
     }
 
-    getAllLinks(): NetworkLink[] {
-        const links: NetworkLink[] = [];
+    getAllLinksInitialState() {
+        const links: Record<
```
Comment (aidemsined) [inline src/data/GraphOnChip.ts:762]: `links` is non descriptive, there shoudl be a mention of what those are recoreded by

File: src/data/GraphOnChip.ts [hunk 7]
```diff
@@ -757,22 +758,93 @@ export default class GraphOnChip {
...
+                    offchipLinkIds: [],
+                    chipId: this.chipId,
+                    saturation: 0,
+                    offchipSaturation: 0,
+                };
+            }
+
             node.links.forEach((link) => {
-                links.push(link);
+                links[node.uid].links[link.uid] = link.generateInitialState();
```
Comment (aidemsined) [inline src/data/GraphOnChip.ts:785]: link[x].links

File: src/data/GraphOnChip.ts [hunk 8]
```diff
@@ -757,22 +758,93 @@ export default class GraphOnChip {
...
     }
 
-    getAllLinks(): NetworkLink[] {
-        const links: NetworkLink[] = [];
+    getAllLinksInitialState() {
+        const links: Record<
+            string,
+            {
+                chipId: number;
+                links: Record<string, LinkState>;
```
Comment (aidemsined) [inline src/data/GraphOnChip.ts:766]: same as above, it shoudl be obvious what the key is and its especially important in a nested record structure

File: src/data/store/selectors/experimentalFeatures.selectors.ts [hunk 9]
```diff
@@ -6,5 +6,5 @@ import { ExperimentalFeaturesState } from 'data/StateTypes';
 import { RootState } from '../createStore';
 
 // eslint-disable-next-line import/prefer-default-export
-export const getExperimentalFeatures = (featureName: keyof ExperimentalFeaturesState) => (state: RootState) =>
+export const getExperimentalFeature = (featureName: keyof ExperimentalFeaturesState) => (state: RootState) =>
```
Comment (aidemsined) [inline src/data/store/selectors/experimentalFeatures.selectors.ts:9]: do we need to keep this if its not being used?

File: src/data/store/slices/linkSaturation.slice.ts [hunk 10]
```diff
@@ -155,58 +150,112 @@ const linkSaturationSlice = createSlice({
...
+    link,
+    totalOpCycles,
+    links,
+    offchipLinkIds,
+}: {
+    link: LinkState;
+    totalOpCycles: number;
+    links: Record<string, LinkState>;
+    offchipLinkIds: string[];
+}) => {
```
Comment (aidemsined) [inline src/data/store/slices/linkSaturation.slice.ts:218]: defined method signature would be useful

File: src/data/store/slices/linkSaturation.slice.ts [hunk 11]
```diff
@@ -155,58 +150,112 @@ const linkSaturationSlice = createSlice({
...
+            });
         });
     });
 };
 
-const recalculateLinkSaturation = (link: LinkState, totalOpCycles: number) => {
-    link.bpc = link.totalDataBytes / totalOpCycles;
-    link.saturation = (link.bpc / link.maxBandwidth) * 100;
+const recalculateLinkSaturation = ({
+    link,
```
Comment (aidemsined) [inline src/data/store/slices/linkSaturation.slice.ts:209]: link and links in method parameters

File: src/data/store/slices/linkSaturation.slice.ts [hunk 12]
```diff
@@ -155,58 +150,112 @@ const linkSaturationSlice = createSlice({
...
+                }
+            });
         });
     });
 };
 
-const recalculateLinkSaturation = (link: LinkState, totalOpCycles: number) => {
-    link.bpc = link.totalDataBytes / totalOpCycles;
-    link.saturation = (link.bpc / link.maxBandwidth) * 100;
+const recalculateLinkSaturation = ({
```
Comment (aidemsined) [inline src/data/store/slices/linkSaturation.slice.ts:208]: the method is clearly doing more than recalculation of saturation, the name shoudl reflect this

Comment (aidemsined) [reply src/data/store/slices/linkSaturation.slice.ts:208]: additionally we _may_ want to consider finally renaming saturation to congestion to bring ui and data model to the same terminology

File: src/data/store/slices/linkSaturation.slice.ts [hunk 13]
```diff
@@ -155,58 +150,112 @@ const linkSaturationSlice = createSlice({
...
+
+                    if (offchipLinkIds.includes(linkId)) {
+                        updatedSaturation.offChipSaturation = Math.max(updatedSaturation.offChipSaturation, saturation);
+                    }
+                }
+
+                return updatedSaturation;
+            },
+            {
+                nodeSaturation: newLinkSaturation,
```
Comment (aidemsined) [inline src/data/store/slices/linkSaturation.slice.ts:237]: node cant be saturated or congested. nodelinks on the other hand can

File: src/data/store/slices/linkSaturation.slice.ts [hunk 14]
```diff
@@ -155,58 +150,112 @@ const linkSaturationSlice = createSlice({
...
     const { PCIBandwidthGBs, CLKHz } = getInitialCLKValues(state);
 
-    Object.values(state.graphs).forEach((graphState) => {
-        Object.values(graphState.links).forEach((linkState: LinkState) => {
-            if (linkState.type === LinkType.PCIE) {
-                linkState.maxBandwidth = PCIBandwidthGBs / CLKHz;
-                recalculateLinkSaturation(linkState, graphState.totalOps);
-            }
+    Object.values(state.linksPerTemporalEpoch).forEach((epochState) => {
+        Object.entries(epochState.linksPerNodeMap).forEach(([nodeUid, { links, offchipLinkIds }]) => {
```
Comment (aidemsined) [inline src/data/store/slices/linkSaturation.slice.ts:187]: `linksPerNodeMap` -> `linksPerNode`? i  actually dislike linksPerNode - there is room for improvent here

File: src/data/GraphOnChip.ts [hunk 15]
```diff
@@ -757,22 +758,93 @@ export default class GraphOnChip {
         return [...nodes];
     }
 
-    getAllLinks(): NetworkLink[] {
-        const links: NetworkLink[] = [];
+    getAllLinksInitialState() {
```
Comment (aidemsined) [inline src/data/GraphOnChip.ts:761]: can we add a signature interface please?

File: src/data/store/selectors/experimentalFeatures.selectors.ts (hunk 9 above)
Comment (madcampos) [reply src/data/store/selectors/experimentalFeatures.selectors.ts:9]: I think for this one, it is not a problem keeping as we will probably use it in the future

File: src/renderer/components/cluster-view/ClusterView.tsx [hunk 16]
```diff
@@ -285,12 +285,11 @@ const ClusterView: FC = () => {
             >
                 {cluster?.chips.map((clusterChip) => {
                     let graphOnChip: GraphOnChip | undefined;
-                    let graphName: string | undefined;
+
                     selectedEpoch.forEach((graph) => {
                         const chipByGraphName = getGraphOnChip(graph.temporalEpoch, graph.chipId);
```
Comment (aidemsined) [inline src/renderer/components/cluster-view/ClusterView.tsx:290]: `chipByGraphName`  is a poor name choice imo. `graphOnChip` should be present in the name but im struggling to finish it. graphOnChipById? :/

Comment (aidemsined) [reply src/renderer/components/cluster-view/ClusterView.tsx:290]: or possibly we could use `graphOnChip` and rename local graphOnChip to `currentGraphOnChip`?

File: src/data/store/selectors/linkSaturation.selectors.ts [hunk 17]
```diff
@@ -2,16 +2,26 @@
...
-    state.linkSaturation.graphs[graphName]?.links || [];
-export const getTotalOpsForGraph = (graphName: string) => (state: RootState) =>
-    state.linkSaturation.graphs[graphName]?.totalOps || 0;
-export const getEpochNormalizedTotalOps = (state: RootState) => state.linkSaturation.epochNormalizedTotalOps;
-export const getEpochAdjustedTotalOps = (state: RootState) => state.linkSaturation.epochAdjustedTotalOps;
+export const getNodeLinksData = (temporalEpoch: number, nodeUid: string) => (state: RootState) =>
+    state.linkSaturation.linksPerTemporalEpoch[temporalEpoch]?.linksPerNodeMap[nodeUid];
+export const getOffchipLinkSaturationForNode = (temporalEpoch: number, nodeUid: string) => (state: RootState) =>
+    state.linkSaturation.linksPerTemporalEpoch[temporalEpoch]?.linksPerNodeMap[nodeUid]?.offchipSaturation ?? 0;
+export const getLinksPerNodeMapForTemporalEpoch = (temporalEpoch: number) => (state: RootState) =>
```
Comment (aidemsined) [inline src/data/store/selectors/linkSaturation.selectors.ts:12]: `getLinksPerNodeForTemporalEpoch`

File: src/data/store/slices/linkSaturation.slice.ts (hunk 13 above)
Comment (madcampos) [reply src/data/store/slices/linkSaturation.slice.ts:237]: I think `nodeTotalLinkSaturation` is a little bit long though... Do you have any suggestion for the name?

File: src/renderer/components/detailed-view-components/DetailedView.tsx [hunk 18]
```diff
@@ -35,9 +36,15 @@ const DetailedView: React.FC<DetailedViewProps> = () => {
...
     const uid = useSelector(getSelectedDetailsViewUID);
     const chipId = useSelector(getSelectedDetailsViewChipId);
-    const graphOnChip = useContext(GraphOnChipContext).getGraphOnChip(temporalEpoch, chipId)[0];
-    const architecture = graphOnChip?.graph.architecture ?? Architecture.NONE;
-    const node = uid ? graphOnChip?.graph.getNode(uid) ?? null : null;
+    const graphOnChip = useContext(GraphOnChipContext).getGraphOnChip(temporalEpoch, chipId ?? -1);
+    const architecture = graphOnChip?.architecture ?? Architecture.NONE;
+    const node = uid ? graphOnChip?.getNode(uid) ?? null : null;
+
+    const linksData = useSelector(getLinksPerNodeMapForTemporalEpoch(temporalEpoch));
```
Comment (aidemsined) [inline src/renderer/components/detailed-view-components/DetailedView.tsx:43]: same, we shoudl limit the scope

File: src/renderer/components/graph-selector/GraphSelector.tsx [hunk 19]
```diff
@@ -20,6 +20,7 @@ interface GraphSelectorProps {
 }
 
 const GraphSelector: FC<GraphSelectorProps> = ({ disabled, label, onSelectGraph, onSelectTemporalEpoch }) => {
+    const IS_TEMPORAL_EPOCH_NAVIGATION_ENABLED = process.env.NODE_ENV === 'development';
```
Comment (aidemsined) [inline src/renderer/components/graph-selector/GraphSelector.tsx:23]: add a TODO: to remove once this is produciton ready?

File: src/data/store/slices/linkSaturation.slice.ts (hunk 13 above)
Comment (aidemsined) [reply src/data/store/slices/linkSaturation.slice.ts:237]: linkSaturation/linkCongestion?

File: src/renderer/components/grid-sidebar/CLKBandwidthControls.tsx [hunk 20]
```diff
@@ -31,37 +31,53 @@ interface DRAMBandwidthControlsProps {}
 
 export const CLKBandwidthControls: FC<DRAMBandwidthControlsProps> = () => {
     const location: Location<LocationState> = useLocation();
-    const { epoch } = location.state;
-    // TODO: use multiple graphs
-    const { graph: graphOnChip, relationship: { name: graphName } = { name: '' } } =
-        useContext(GraphOnChipContext).getGraphOnChip(epoch)[0] ?? {};
+    const { epoch, chipId } = location.state;
+
```
Comment (aidemsined) [inline src/renderer/components/grid-sidebar/CLKBandwidthControls.tsx:35]:
``` js
const graphOnChipList = useContext(GraphOnChipContext).getGraphOnChipListForTemporalEpoch(epoch);
const graphOnchip = chipId!== undefined ? graphOnChipList[chipId]: undefined;
```
?

File: src/renderer/components/grid-sidebar/CongestionControls.tsx [hunk 21]
```diff
@@ -32,18 +32,32 @@ import type { LocationState } from '../../../data/StateTypes';
 
 export const CongestionControls: FC = () => {
     const location: Location<LocationState> = useLocation();
-    const { epoch } = location.state;
-    // TODO: use multiple graphs
-    const graphOnChip = useContext(GraphOnChipContext).getGraphOnChip(epoch)[0]?.graph;
+    const { epoch, chipId } = location.state;
+
+    let graphOnChipList = useContext(GraphOnChipContext).getGraphOnChipListForTemporalEpoch(epoch);
```
Comment (aidemsined) [inline src/renderer/components/grid-sidebar/CongestionControls.tsx:37]: same as CLKBandwidthControls

File: src/renderer/components/grid-sidebar/ModelControls.tsx [hunk 22]
```diff
@@ -17,26 +17,38 @@ import type { LocationState } from '../../../data/StateTypes';
 
 const ModelControls: FC = () => {
     const location: Location<LocationState> = useLocation();
-    const { epoch } = location.state;
-    // TODO: use multiple graphs
-    const graphOnChip = useContext(GraphOnChipContext).getGraphOnChip(epoch)[0]?.graph;
+    const { epoch, chipId } = location.state;
+
+    let graphOnChipList = useContext(GraphOnChipContext).getGraphOnChipListForTemporalEpoch(epoch);
```
Comment (aidemsined) [inline src/renderer/components/grid-sidebar/ModelControls.tsx:22]: same as CLKBandwidthControls

File: src/renderer/components/properties-panel/ComputeNodesPropertiesTab.tsx [hunk 23]
```diff
@@ -381,12 +382,12 @@ const ComputeNodesPropertiesTab = ({
...
             <div className='properties-list'>
                 <div className='properties-panel-nodes'>
-                    {selectedNodes.map(({ node, graphName }) => (
+                    {selectedNodes.map((node) => (
                         <ComputeNodePropertiesCard
                             key={node?.uid}
                             node={node}
                             temporalEpoch={epoch}
-                            graphName={graphName}
+                            allLinksState={allLinksState}
```
Comment (aidemsined) [inline src/renderer/components/properties-panel/ComputeNodesPropertiesTab.tsx:390]: we dont need all link states passed here, we can just grab what we need maybe?

File: src/renderer/components/GridRender.tsx [hunk 24]
```diff
@@ -25,7 +25,16 @@ export default function GridRender() {
     const location: Location<LocationState> = useLocation();
     const { chipId, epoch } = location.state;
 
-    const graphOnChipList = useContext(GraphOnChipContext).getGraphOnChip(epoch, chipId);
+    let graphOnChipList = useContext(GraphOnChipContext).getGraphOnChipListForTemporalEpoch(epoch);
```
Comment (aidemsined) [inline src/renderer/components/GridRender.tsx:28]: same as CLKBandwidthControls

File: src/renderer/components/cluster-view/EthPipeRenderer.tsx [hunk 25]
```diff
@@ -64,7 +64,7 @@ const EthPipeRenderer: FC<EthPipeRendererProps> = ({
 
     const showLinkSaturation = useSelector(getShowLinkSaturation);
     const linkSaturationTreshold = useSelector(getLinkSaturation);
-    const linksData = useSelector(getAllLinksForGraph(graphName || ''));
+    const linksData = useSelector(getLinksPerNodeMapForTemporalEpoch(temporalEpoch));
```
[the remark this answers is not in the document]
Comment (madcampos) [reply src/renderer/components/cluster-view/EthPipeRenderer.tsx:67]: Yes and no... We can limit, but it results in partial or missing datam specially for things like details view.

I think there is room for improvement in a subsequent PR.

File: src/renderer/components/PipeInfoDialog.tsx [hunk 26]
```diff
@@ -25,24 +25,36 @@ export interface PipeInfoDialogProps {
  */
 const PipeInfoDialog: FC<PropsWithChildren<PipeInfoDialogProps>> = ({ children, pipeId, hide, onEnter, onLeave }) => {
     const location: Location<LocationState> = useLocation();
-    const { epoch } = location.state;
-    // TODO: use multiple graphs
-    const graphOnChip = useContext(GraphOnChipContext).getGraphOnChip(epoch)[0]?.graph;
+    const { epoch, chipId } = location.state;
+    let graphOnChipList = useContext(GraphOnChipContext).getGraphOnChipListForTemporalEpoch(epoch);
```
Comment (aidemsined) [inline src/renderer/components/PipeInfoDialog.tsx:29]: ditto

File: src/data/store/selectors/experimentalFeatures.selectors.ts (hunk 9 above)
Comment (aidemsined) [reply src/data/store/selectors/experimentalFeatures.selectors.ts:9]: 👍

File: src/renderer/components/detailed-view-components/DetailedView.tsx (hunk 18 above)
Comment (aidemsined) [reply src/renderer/components/detailed-view-components/DetailedView.tsx:43]: probably todo for now

File: src/renderer/components/properties-panel/ComputeNodesPropertiesTab.tsx (hunk 23 above)
Comment (madcampos) [reply src/renderer/components/properties-panel/ComputeNodesPropertiesTab.tsx:390]: Yes, this is not the optimal solution, I'll add a TODO, but maybe fix it on another PR.

File: src/data/store/slices/linkSaturation.slice.ts (hunk 13 above)
Comment (madcampos) [reply src/data/store/slices/linkSaturation.slice.ts:237]: But it is for all links on the node, pre calculated

File: src/renderer/components/cluster-view/EthPipeRenderer.tsx (hunk 25 above)
[the remark this answers is not in the document]
Comment (madcampos) [reply src/renderer/components/cluster-view/EthPipeRenderer.tsx:67]: This is the bottleneck for details view

File: src/data/store/slices/linkSaturation.slice.ts [hunk 27]
```diff
@@ -155,58 +160,123 @@ const linkSaturationSlice = createSlice({
...
+                });
+            },
+        );
     });
 };
 
-const recalculateLinkSaturation = (link: LinkState, totalOpCycles: number) => {
-    link.bpc = link.totalDataBytes / totalOpCycles;
-    link.saturation = (link.bpc / link.maxBandwidth) * 100;
+interface RecalculateLinkStaturationArguments {
```
Comment (aidemsined) [inline src/data/store/slices/linkSaturation.slice.ts:224]: `LinkStaturationMetrics`?

File: src/data/GraphOnChipContext.tsx [hunk 28]
```diff
@@ -85,46 +89,54 @@ const GraphOnChipProvider: React.FC<{ children: ReactNode }> = ({ children }) =>
...
+                (graph) => graph.temporalEpoch === temporalEpoch && graph.chipId === chipId,
+            );
 
-            return graphs;
+            return state.graphOnChipList[graphRelationship?.name ?? ''];
         },
         [state.graphs, state.graphOnChipList],
     );
 
     const reset = useCallback(() => {
```
Comment (aidemsined) [inline src/data/GraphOnChipContext.tsx:133]: `resetGraphOnChipState`?

File: src/renderer/components/properties-panel/ComputeNodesPropertiesTab.tsx [hunk 29]
```diff
@@ -341,7 +342,7 @@ const ComputeNodePropertiesCard = ({ node, temporalEpoch, graphName }: ComputeNo
                 <div className='node-links-wrap'>
                     <h4>Links</h4>
                     {node.getNOCLinksForNode().map((link: NOCLink) => (
-                        <LinkDetails key={link.name} link={link} graphName={graphName} showEmpty />
+                        <LinkDetails key={link.name} link={link} linkState={allLinksState[link.uid]} showEmpty />
```
Comment (aidemsined) [inline src/renderer/components/properties-panel/ComputeNodesPropertiesTab.tsx:345]: we have a method to only get a single node links, lets use that

Comment (madcampos) [reply src/renderer/components/properties-panel/ComputeNodesPropertiesTab.tsx:345]: I'll fix that afterwards, along with the other node links.

But this is inside of a loop and requires state, so we can't call a selector directly.

Comment (aidemsined) [reply src/renderer/components/properties-panel/ComputeNodesPropertiesTab.tsx:345]: you can grab a selector from within a component, you dont have to pass it in

Comment (aidemsined) [reply src/renderer/components/properties-panel/ComputeNodesPropertiesTab.tsx:345]: while we could fix this in the next round this is an easy one to fix and an easy one to miss later

**Review by aidemsined — APPROVED** [review]

**Pull request merged**